r1cs/gadgets/std/src/bits/macros.rs

```diff
@@ -0,0 +1,2532 @@
+                        let mut multi_eq = MultiEq::new(&mut cs);
+                        self.sub(multi_eq.ns(|| "a - b mod 2^n"), other)?
+                    };
+
+                    let mut delta_lc = LinearCombination::zero();
+                    let mut coeff = ConstraintF::one();
+                    let mut all_constants = true;
+                    for ((self_bit, other_bit), diff_bit) in self.bits.iter().zip(other.bits.iter()).zip(diff_var.bits.iter()) {
+                        delta_lc = delta_lc + &self_bit.lc(CS::one(), coeff);
+                        delta_lc = delta_lc - &other_bit.lc(CS::one(), coeff);
+                        delta_lc = delta_lc - &diff_bit.lc(CS::one(), coeff);
+
+                        all_constants &= self_bit.is_constant() && other_bit.is_constant();
+
+                        coeff.double_in_place();
+                    }
+
+                    let (diff_val, is_underflowing, delta) = match (self.get_value(), other.get_value()) {
+                        (Some(value1), Some(value2)) => {
+                            let (diff, underflow) = value1.overflowing_sub(value2);
+                            // compute delta = self - other - diff in the field,
+                            // where diff = self - other mod 2^$bit_size
+                            let self_in_field = ConstraintF::from(value1);
+                            let other_in_field = ConstraintF::from(value2);
+                            let diff_in_field = ConstraintF::from(diff);
+                            let delta = self_in_field - other_in_field - diff_in_field;
+                            (Some(diff), Some(underflow), Some(delta))
+                        },
+                        _ => (None, None, None),
+                    };
+
+                    if all_constants && diff_val.is_some() {
+                        return Ok(Boolean::constant(is_underflowing.unwrap()))
+                    }
+
+                    // ToDo: It should not be necessary to allocate it as a Boolean gadget,
+                    // can be done when a Boolean::from(FieldGadget) will be implemented
+                    let is_smaller = Boolean::alloc(cs.ns(|| "alloc result"), || is_underflowing.ok_or(SynthesisError::AssignmentMissing))?;
+
+                    let inv = delta.map(|delta| {
+                        match delta.inverse() {
+                            Some(inv) => inv,
+                            None => ConstraintF::one(), // delta is 0, so we can set any value
+                        }
+                    });
+
+                    let inv_var = FpGadget::<ConstraintF>::alloc(cs.ns(|| "alloc inv"), || inv.ok_or(SynthesisError::AssignmentMissing))?;
+
+                    // enforce constraints:
+                    // (1 - is_smaller) * delta_lc = 0 enforces that is_smaller == 1 when delta != 0, i.e., when a < b
+                    // inv * delta_lc = is_smaller enforces that is_smaller == 0 when delta == 0, i.e., when b >= a
```

Comment:
Comment is wrong here, should be `a >= b`

Reply:
Corrected, thanks